jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4296,36 +4296,108 @@ ConstantPool getConstantPool() {
  */
 private native Class<?>[] getNestMembersImpl();
 
+/* Determines if 'currentClassLoader' is the same or a child of the requestedLoader. */
+private static boolean doesClassLoaderDescendFrom(ClassLoader currentLoader, ClassLoader requestedLoader) {
+	if (requestedLoader == null) {
+		/* Bootstrap loader is parent of everyone */
+		return true;
+	}
+	if (currentLoader != requestedLoader) {
+		while (currentLoader != null) {
+			if (currentLoader == requestedLoader) {
+				return true;
+			}
+			currentLoader = currentLoader.getParent();
+		}
+		return false;
+	}
+	return true;
+}
+
 /**
  * Answers the host class of the receiver's nest.
  *
  * @return		the host class of the receiver.
  */
-public Class<?> getNestHost() {
-	return getNestHostImpl();
+public Class<?> getNestHost() throws SecurityException {
+	Class<?> nestHost = getNestHostImpl();
+	
+	/* The specification requires that if:
+	 *    - the returned class is not the current class
+	 *    - a security manager is present
+	 *    - the caller's class loader is not the same or an ancestor of the returned class
+	 *    - s.checkPackageAccess() disallows access to the package of the returned class
+	 * then throw a SecurityException.
+	 */
+	if (nestHost != this) {
+		SecurityManager securityManager = System.getSecurityManager();
+		if (nestHost == null) {
+			nestHost = this;
+		} else if (securityManager != null) {
+			ClassLoader callerClassLoader = ClassLoader.getCallerClassLoader();
+			ClassLoader nestHostClassLoader = nestHost.getClassLoader();			
+			if (!doesClassLoaderDescendFrom(nestHostClassLoader, callerClassLoader)) {
+				String nestHostPackageName = nestHost.getPackageName();
+				if (nestHostPackageName != null) {
+					securityManager.checkPackageAccess(nestHostPackageName);
+				}
+			}
+		}
+	}
+
+	return nestHost;
 }
 
 /**
  * Returns true if the class passed has the same nest top as this class.
- * The list of nest members in the classfile is permitted to contain
- * duplicates, or to explicitly include the nest host. It is not required
- * that an implementation of this method removes these duplicates.
  * 
  * @param		that		The class to compare
  * @return		true if class is a nestmate of this class; false otherwise.
  *
  */
 public boolean isNestmateOf(Class<?> that) {
-	return this.getNestHost() == that.getNestHost();
+	return this.getNestHostImpl() == that.getNestHostImpl();
```

Comment:
@DanHeidinga 
I find it odd that `isNestmateOf` can't throw a security exception but `getNestHost` does.
From the spec `Two classes are nestmates if they have the same nest host .`

If a caller doesn't have permission query the nesthost, should it be allowed to test whether the nest host is equivalent to another classes nesthost?